hudi-spark-datasource/hudi-spark3-common/src/main/scala/org/apache/spark/sql/HoodieSpark3CatalystPlanUtils.scala

```diff
@@ -52,8 +57,56 @@ abstract class HoodieSpark3CatalystPlanUtils extends HoodieCatalystPlansUtils {
     }
   }
 
-  override def toTableIdentifier(relation: UnresolvedRelation): TableIdentifier = {
-    relation.multipartIdentifier.asTableIdentifier
+  override def resolve(spark: SparkSession, relation: UnresolvedRelation): Option[CatalogTable] = {
```

Comment:
Can you please elaborate why we need to have this custom resolution rule even though there's Spark's default `ResolveRelations` one?

Reply:
sure. see the user case: https://github.com/apache/hudi/issues/6223.
If we use the three part `catalog.database.table` in spark3.x like , we need the ability to parse these identifiers.
And, the reason that Spark's default `ResolveRelations` can't work is, the `UnresolvedRelation` hasn't resolved when apply the cases/rules in `HoodieAnalysis`.

Reply:
the test I have added without these changes is same with this user's case: https://github.com/apache/hudi/issues/6223#issuecomment-1198952333.
![image](https://user-images.githubusercontent.com/10036681/183798356-c0e0b983-7abf-4fcd-89e0-015d287f9c85.png)

Reply:
But my question is why is this not being handled properly by Spark?

Reply:
Please check my comment here:
https://github.com/apache/hudi/pull/6264#discussion_r941962972

Reply:
understand. It's really a question that deserve to be dived.
But a same question: why we also need to resolve `UnresolvedRelation` or `UnresolvedTable` in inside of hoodie even in spark or where no such problem need to be solved.

Reply:
@alexeykudinkin both `UnresolvedRelation` and `UnresolvedTable` call `resolveOperatorsUpWithPruning` to make itself and it's children to apply the Rule.
![image](https://user-images.githubusercontent.com/10036681/183813122-93e16d66-15d1-4307-ae1b-149d67b0e000.png)

But for the sql `insert into target_table select col1, col2 from source_table`, only `select col1, col2 from source_table` as the query of `InsertIntoStatement` is it's child, `target_table` as the table of `InsertIntoStatement` is not. So the table LogicalPlan can't be applied by `ResolvedRelation` or `ResolvedTable`.
![image](https://user-images.githubusercontent.com/10036681/183814152-b043b4b0-013b-4ddd-8c52-b7f8f51aeeee.png)

Reply:
@YannByron i don't think it is the case simply b/c if that would be the case, then `InsertIntoStatement` wouldn't be resolved by `ResolveRelations` rule which it most certainly does (take a look at `FindDataSourceTable` in [here](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala#L282))

Reply:
Let me explain this in detail.
1) Hudi injects `HoodieResolveReferences` into the `Resolution` batch in Analyzer.
2) The `Resolution` batch will have these Rule in order: `ResolveInsertInto`, `ResolveRelations`, `ResolveTables`, ... , `FindDataSourceTable`, ..., `HoodieResolveReferences`.
3) the sql `insert into target_table select * from source_table` will be analyzed by the rules above in order.
**In the first iterator** over Analyzer's `Resolution` batch: 
- `ResolveInsertInto` isn't used.
- `ResolveRelations` is used to resolve the child of `InsertIntoStatement` (that is query).Here the target_table doesn't be resolved.
- `FindDataSourceTable` is not used, b/c the first argument (`table`) of `InsertIntoStatement` is still an `UnresolvedRelation`, not an `UnresolvedCatalogRelation`.
- then `HoodieResolveReferences` will be applied. In it, hudi gets the `table` object (an `UnresolvedRelation` for now) from `InsertIntoStatement`, and want to determine it is a hudi table or not according to this. It will call the origin logical that try to transform `UnresolvedRelation` to `TableIdentifier` directly. But if it's a three-part format, an exception will be thrown like this case.

But, for normal spark which do not have this `HoodieResolveReferences` Rule, the `table` object of `InsertIntoStatement` will be resolved by `ResolveTables` **in the second iterator** over Analyzer's `Resolution` batch. That's the reason why spark can resolve this correctly.

I think the root cause is that hudi injects `HoodieResolveReferences` into `Resolution`, not into `Post-Hoc Resolution`. If so, maybe hudi never need to resolve `UnresolvedXXXX` by itself. But it's a big changes, need an other issue to trace.
So this pr is the controllable solution to solve the user case that only do a few changes in current hudi architecture.

Reply:
@YannByron thanks for detailed elaboration! Appreciate that!

The point i’m trying to make is that Spark _should be_ resolving the relations, NOT Hudi — we should not take away functionality from Spark that we don’t customize, extend or fix. In this case we are not doing either of these things so we should leave resolving to Spark as it is.

My suggestions are basically following:

1. Avoid copying any resolution logic in your PR, and instead
2. Fix things that fail individually:

    - `isHoodieTable` should NOT be invoked on `UnresolvedRelation`. Instead we should add conditional checking whether target relation in `InsertIntoStatement`  has been fully resolved by Spark.
    - Same goes for `TimeTravelRelation` — it’s currently actually implemented incorrectly. `TimeTravelRelation` is implemented as `Command` , while in reality it should just be either `LeafNode` or `UnaryNode` (in which case it would be resolved by Spark automatically)

Reply:
> Moreover this part is not even affected by your changes. Your changes only affect Time Travel queries, and not Insert Into statement.

not just insert statement. If you run `select * from spark_catalog.default.hudi_tbl timestamp as of 111111` in the master branch, it also throws the same exception. This pr can also solve this case.

Reply:
> My suggestions are basically following: 1. 2.

your suggestions can also work. Still inject the rules to  `Resolution` batch but force to ask all the relation is resolved.
but for the time travel case you also mention, we have to resolve the `UnresolvedRelation` inside of Hudi by this resolution logic in this pr.

Reply:
You're right `TimeTravelRelation` part will actually have to stay since it's an ephemeral construct we ultimately replace w/ `LogicalRelation`

Reply:
> I suggest to create an other pr to trace. And this pr just makes the identifier with catalog can work in hudi.

I understand that it's not ideal to make you revisit the approach, but i don't think it makes sense to introduce something in this PR that we might need to be reverting soon after.